### PR TITLE
Incremental compilation & hotreloading tests

### PR DESCRIPTION
**Pull request by baszalmstra**
This commit restructures the compiler crates to better support incremental compilation. The compiler `--watch` mode is now incremental, reusing the state from previous compilations.

**Todo:**

- [x] Implement `Driver`
- [x] Make `--watch` mode incremental with `Driver`
- [x] Add a test to check `--watch` mode works
- [x] Add a runtime test to verify hotloading

### CODE REVIEW

**Comment by baszalmstra** [comment]
[Add incremental compilation to compiler daemon by properly using salsa](https://trello.com/c/pVl2KU35/51-add-incremental-compilation-to-compiler-daemon-by-properly-using-salsa)

**Comment by baszalmstra** [comment]
[Create tests for hot reloading functionality](https://trello.com/c/4r3xDxWh/59-create-tests-for-hot-reloading-functionality)

File: crates/mun_compiler/src/diagnostics/emit.rs [hunk 1]
```diff
@@ -42,7 +42,7 @@ impl Emit for Diagnostic {
         writer.set_color(&error)?;
         write!(writer, "error")?;
 
-        // Write diagnostic message
+        // Write diagnostics message
```
Comment (Wodann) [inline crates/mun_compiler/src/diagnostics/emit.rs]: singular is diagnostic, plural is diagnostics.

File: crates/mun_compiler/src/driver.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+pub const WORKSPACE: SourceRootId = SourceRootId(0);
+
+#[derive(Debug)]
+pub struct Driver {
+    db: CompilerDatabase,
+    out_dir: Option<PathBuf>,
+}
+
+impl Driver {
+    /// Construct a driver with a specific configuration.
```
Comment (Wodann) [inline crates/mun_compiler/src/driver.rs]:
```suggestion
    /// Constructs a driver with a specific configuration.
```

File: crates/mun_compiler/src/driver.rs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    pub fn with_config(config: Config) -> Self {
+        let mut driver = Driver {
+            db: CompilerDatabase::new(),
+            out_dir: None,
+        };
+
+        // Move relevant configuration into the database
+        driver.db.set_target(config.target);
```
Comment (Wodann) [inline crates/mun_compiler/src/driver.rs]: Would a single `set_config` be an option? That would simplify adding options and reduce *failure points*.

File: crates/mun_compiler/src/driver.rs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    pub fn diagnostics(&self) -> Vec<Diagnostic> {
+        self.db
+            .source_root(WORKSPACE)
+            .files()
+            .map(|f| diagnostics(&self.db, f))
+            .flatten()
+            .collect()
+    }
+
+    /// Emits all diagnostic messages currently in the database; returns true if errors where
```
Comment (Wodann) [inline crates/mun_compiler/src/driver.rs]:
```suggestion
    /// Emits all diagnostic messages currently in the database; returns true if errors were
```

File: crates/mun_runtime/src/test.rs [hunk 5]
```diff
@@ -1,95 +1,143 @@
...
+        TestDriver {
+            _temp_dir: temp_dir,
+            driver,
+            out_path,
+            file_id,
+            runtime,
+        }
+    }
+
+    /// Updates the text of the Mun source and ensure the generated assembly has been reloaded.
```
Comment (Wodann) [inline crates/mun_runtime/src/test.rs]:
```suggestion
    /// Updates the text of the Mun source and ensures that the generated assembly has been reloaded.
```

File: crates/mun_runtime/src/test.rs [hunk 6]
```diff
@@ -1,95 +1,143 @@
...
+            if now - start_time > std::time::Duration::from_secs(10) {
+                panic!("runtime did not update after recompilation within 10secs");
+            } else {
+                sleep(Duration::from_millis(1));
+            }
+        }
+    }
+
+    /// Calls the function with the specified name
+    fn invoke0<T: Reflection>(&mut self, fn_name: &str) -> T {
```
Comment (Wodann) [inline crates/mun_runtime/src/test.rs]: Why don't you add a `runtime_mut()` function to `TestDriver`? That way you don't have to duplicate the `invoke_fn` logic.

File: crates/mun_compiler/src/diagnostics/emit.rs (hunk 1 above)
Comment (baszalmstra) [reply crates/mun_compiler/src/diagnostics/emit.rs]: I think I renamed a field which also renamed comments ...

File: crates/mun_compiler/src/driver.rs (hunk 3 above)
Comment (baszalmstra) [reply crates/mun_compiler/src/driver.rs]: Do you mean on the `Driver` or on the `CompilerDatabase`? There are two considerations

1. The `Config` can contain options for both and/or either the `Driver` and the `Database`.
2. Constructing a `Driver` or a `CompilerDatabase` without configuration doesnt make sense.

File: crates/mun_runtime/src/lib.rs [hunk 7]
```diff
@@ -131,7 +131,7 @@ impl Runtime {
         assembly.link(&self.dispatch_table)?;
 
         self.watcher
-            .watch(library_path.clone(), RecursiveMode::NonRecursive)?;
+            .watch(library_path.parent().unwrap(), RecursiveMode::NonRecursive)?;
```
[the remark this answers is not in the document]
Comment (baszalmstra) [reply crates/mun_runtime/src/lib.rs]: The first line of the method canonicalizes the path so it should be an existing absolute path.

File: crates/mun_runtime/src/test.rs (hunk 6 above)
Comment (baszalmstra) [reply crates/mun_runtime/src/test.rs]: Because this is shorter which I find much clearer.

File: crates/mun_compiler/src/driver.rs (hunk 3 above)
Comment (Wodann) [reply crates/mun_compiler/src/driver.rs]: The compiler database. Now you need to manually set all of the different config fields.

File: crates/mun_runtime/src/test.rs (hunk 6 above)
Comment (Wodann) [reply crates/mun_runtime/src/test.rs]: The trade-off is that a lot more code is added overall.

**Pull request merged**